### PR TITLE
Codepen Links with Underscore Usernames Flagged as Invalid

### LINKED ISSUE #2981: Invalid CodePen URL for usernames containing an underscore
**Describe the bug**
When trying to create a new post which contains a CodePen link, the editor will state that the CodePen link is invalid if the username within the CodePen url provided contains an underscore.

**To Reproduce**

_Example users with underscores:_
https://codepen.io/George_1/
https://codepen.io/cobra_winfrey/

_Example urls with underscores:_
https://codepen.io/George_1/pen/jrBXPA
https://codepen.io/cobra_winfrey/pen/yXVMJP

_Example CodePen Dev.To links with underscores:_
`{% codepen https://codepen.io/cobra_winfrey/pen/yXVMJP %}`
`{% codepen https://codepen.io/George_1/pen/jrBXPA %}`

1. Create a new post
2. Populate the post as you would expect
3. Add a CodePen link to the post containing one of the example urls provided above or any pen from a user on CodePen with an underscore in the name
4. Try and preview the post
5. Try and save changes

**Expected behavior**
Upon preview I would expect to see a working preview of the post with a working CodePen.
Upon save changes I would expect my post to be saved as normal.

**Screenshots**
image

**Desktop (please complete the following information):**

- OS: Arch Linux, up to date as of time of posting
- Browser: Chrome 
- Version: 73

### PR DESCRIPTION
**Pull request by aturingmachine**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [ ] Feature
- [X] Bug Fix
- [ ] Documentation Update

## Description
Updates the liquid tag regex to allow underscores. Also adds a test using an underscore in the username section of the url to verify changes work.

## Related Tickets & Documents
Closes #2981 
## Mobile & Desktop Screenshots/Recordings (if there are UI changes)
N/A
## Added to documentation?

- [ ] docs.dev.to
- [ ] readme
- [X] no documentation needed

### CODE REVIEW

File: app/liquid_tags/codepen_tag.rb [hunk 1]
```diff
@@ -47,7 +47,7 @@ def parse_link(link)
   def valid_link?(link)
     link_no_space = link.delete(" ")
     (link_no_space =~
-      /^(http|https):\/\/(codepen\.io|codepen\.io\/team)\/[a-zA-Z0-9\-]{1,20}\/pen\/([a-zA-Z]{5,7})\/{0,1}\z/)&.zero?
+      /^(http|https):\/\/(codepen\.io|codepen\.io\/team)\/[a-zA-Z0-9_\-]{1,20}\/pen\/([a-zA-Z]{5,7})\/{0,1}\z/)&.zero?
```
Comment (ChrisBertrand) [inline app/liquid_tags/codepen_tag.rb]: this only allows the username to be upto 20 characters long, in my case designpuddle-the-stylish fails the check. Why was this limit set?

Comment (Zhao-Andy) [reply app/liquid_tags/codepen_tag.rb]: I believe they have a 20 character limit for usernames now, but my guess is that they added that limit and that there are some accounts that are grandfathered in. Our limit is loosely based on those assumptions, so a PR to adjust it is more than welcome!

Comment (ChrisBertrand) [reply app/liquid_tags/codepen_tag.rb]: I created a new account today, and it assigned me a username with 24 characters. I've managed to change the username so I can embed on dev.to but thought I would mention it. I'm happy to create the PR if needed, what would the limit be raised to in that instance?

Comment (Zhao-Andy) [reply app/liquid_tags/codepen_tag.rb]: I think 30 characters is a good start. If you could reference this issue that'd be great: https://github.com/thepracticaldev/dev.to/issues/3487

**Pull request merged**